### PR TITLE
Add workflow.GetInfo(ctx).GetCurrentHistoryLength()

### PR DESCRIPTION
**Pull request by cretz**
## What was changed

Added `workflow.GetInfo(ctx).GetCurrentHistoryLength()` and a helper to set during unit tests

## Why?

So people can use it to make continue-as-new determinations

## Checklist

1. Part of #628

### CODE REVIEW

File: test/integration_test.go [hunk 1]
```diff
@@ -2425,6 +2425,32 @@ func (ts *IntegrationTestSuite) registerNamespace() {
 	}
 }
 
+func (ts *IntegrationTestSuite) TestHistoryLength() {
```
Comment (bergundy) [inline test/integration_test.go:2428]: To get some extra coverage, I'd evict this workflow from cache (fail a task?) to verify that the history length is correct during replay too.

**Review by bergundy — APPROVED** [review]
Approved with single comment

File: test/integration_test.go (hunk 1 above)
Comment (cretz) [reply test/integration_test.go:2428]: Maybe I can complete the workflow then run a query for the same value I'm returning

Comment (cretz) [reply test/integration_test.go:2428]: Note, our integration tests run in no-cache mode too

Comment (bergundy) [reply test/integration_test.go:2428]: That should be enough then, I'd just maybe mention that we get the coverage from that

File: internal/internal_event_handlers.go [hunk 2]
```diff
@@ -826,6 +826,8 @@ func (weh *workflowExecutionEventHandlerImpl) ProcessEvent(
 	case enumspb.EVENT_TYPE_WORKFLOW_TASK_STARTED:
 		// Set replay clock.
 		weh.SetCurrentReplayTime(common.TimeValue(event.GetEventTime()))
+		// Set history length as this event's ID
+		weh.workflowInfo.currentHistoryLength = int(event.EventId)
```
Comment (Sushisource) [inline internal/internal_event_handlers.go:830]: I don't know if this is sufficient. Based on what's going on here: https://cs.github.com/temporalio/sdk-go/blob/2f76f4de20f0171cb67a6881853c6969f2324761/internal/internal_task_handlers.go#L910

Event ids can be processed out-of-order when looking at markers, which makes sense, but also means this line should probably be more like:
`weh.workflowInfo.currentHistoryLength = max(weh.workflowInfo.currentHistoryLength, int(event.EventId))

File: test/integration_test.go (hunk 1 above)
Comment (Sushisource) [reply test/integration_test.go:2428]: Definitely add some tests using local activities with and without replay too, per the ordering thing I mention in my other comment

File: internal/internal_event_handlers.go (hunk 2 above)
Comment (cretz) [reply internal/internal_event_handlers.go:830]: This is just the event ID of workflow task started. While we do process change the order of history during processing, I believe we only do it one task at a time. So at the top of that loop iteration from your link there is `reorderedHistory.NextCommandEvents()`. Arguably I could just put this value right after we obtain that, but I think the effect is the same.

I may have to re-read the code to confirm, it is a bit hard to follow.

File: internal/internal_workflow_testsuite.go [hunk 3]
```diff
@@ -338,6 +338,10 @@ func (env *testWorkflowEnvironmentImpl) setStartTime(startTime time.Time) {
 	env.workflowInfo.WorkflowStartTime = env.mockClock.Now()
 }
 
+func (env *testWorkflowEnvironmentImpl) setCurrentHistoryLength(length int) {
+	env.workflowInfo.currentHistoryLength = length
+}
```
[the remark this answers is not in the document]
Comment (Sushisource) [reply internal/internal_workflow_testsuite.go:343]: Got it

File: test/integration_test.go (hunk 1 above)
Comment (cretz) [reply test/integration_test.go:2429]: Updated the integration test to also use local activities and added comment making it clear we use the zero-cache integration test to confirm replay (and have also manually confirmed that the zero-cache does replay properly).
